server/src/actors/project.rs

```diff
@@ -894,73 +903,94 @@ impl ProjectCache {
         // reinsert them at the end, as they are now receiving an updated timestamp. Then,
         // batch-insert all new projects with the new timestamp.
         self.updates
-            .retain(|update| !batch.contains_key(&update.project_id));
+            .retain(|update| !channels.contains_key(&update.project_id));
         self.updates
-            .extend(batch_ids.iter().copied().map(ProjectUpdate::new));
+            .extend(projects.iter().copied().map(ProjectUpdate::new));
 
         metric!(timer("project_state.eviction.duration") = eviction_start.elapsed());
-        metric!(histogram("project_state.request") = batch.len() as u64);
         metric!(histogram("project_state.pending") = self.state_channels.len() as u64);
 
         log::debug!(
             "updating project states for {}/{} projects (attempt {})",
-            batch.len(),
-            batch.len() + self.state_channels.len(),
+            channels.len(),
+            channels.len() + self.state_channels.len(),
             self.backoff.attempt(),
         );
 
-        let request = GetProjectStates {
-            projects: batch_ids,
-            #[cfg(feature = "processing")]
-            full_config: self.config.processing_enabled(),
-        };
-
-        // count number of http requests for project states
-        metric!(counter("project_state.request") += 1);
-        let request_start = Instant::now();
+        let requests: Vec<_> = channels
+            .into_iter()
+            .chunks(batch_size)
+            .into_iter()
+            .map(|channels_batch| {
+                let channels_batch: HashMap<_, _> = channels_batch.collect();
+                log::debug!("sending request of size {}", channels_batch.len());
+                metric!(histogram("project_state.request") = channels_batch.len() as u64);
+
+                let request = GetProjectStates {
+                    projects: channels_batch.keys().copied().collect(),
+                    #[cfg(feature = "processing")]
+                    full_config: self.config.processing_enabled(),
+                };
+
+                // count number of http requests for project states
+                metric!(counter("project_state.request") += 1);
+                let request_start = Instant::now();
+
+                self.upstream
+                    .send(SendQuery(request))
+                    .map_err(ProjectError::ScheduleFailed)
+                    .map(move |response| (channels_batch, request_start, response))
+            })
+            .collect();
 
-        self.upstream
-            .send(SendQuery(request))
-            .map_err(ProjectError::ScheduleFailed)
+        future::join_all(requests)
```

Comment:
Can you please add a comment about the behavior of `join_all` here? To double check: If all requests run to completion (even with failures), unless the mailbox of the `Upstream` actor is full.

Reply:
done